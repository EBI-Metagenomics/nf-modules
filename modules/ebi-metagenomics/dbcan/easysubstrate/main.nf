process DBCAN {
    tag "${meta.id}"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/dbcan:5.1.2--pyhdfd78af_0'
        : 'biocontainers/dbcan:5.1.2--pyhdfd78af_0'}"

    input:
    tuple val(meta), path(fasta)
    tuple val(meta2), path(gff) // only used for cluster with a proteins file
    tuple path(dbcan_db, stageAs: "dbcan_db"), val(db_version)
    val(mode)

    output:
    tuple val(meta), path("results/${prefix}_overview.tsv.gz"),                             emit: overview_txt
    tuple val(meta), path("results/${prefix}_dbcan_hmm_results.tsv.gz"),                    emit: dbhmm_output_tsv
    tuple val(meta), path("results/${prefix}_dbcansub_hmm_results.tsv.gz"),                 emit: dbsub_output_tsv
    tuple val(meta), path("results/${prefix}_diamond.out.gz"),                              emit: diamond_output
    tuple val(meta), path("results/${prefix}_uniinput.faa.gz"),             optional: true, emit: uniinput_faa
    tuple val(meta), path("results/${prefix}_cgc.gff.gz"),                  optional: true, emit: cgc_gff
    tuple val(meta), path("results/${prefix}_cgc_standard_out.tsv.gz"),     optional: true, emit: cgc_standard_tsv
    tuple val(meta), path("results/${prefix}_substrate_prediction.tsv.gz"), optional: true, emit: substrate_prediction_tsv
    tuple val(meta), path("results/synteny_pdf/*-syntenic.pdf.gz"),         optional: true, emit: synteny_pdfs
    path "versions.yml",                                                                    emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"

    def is_fasta_compressed = fasta.getExtension() == "gz"
    def fasta_name = is_fasta_compressed ? fasta.getBaseName() : fasta

    def is_gff_compressed = gff.getExtension() == "gz"
    def gff_name = is_gff_compressed ? gff.getBaseName() : gff
    """
    if [ "${is_fasta_compressed}" == "true" ]; then
        gzip -c -d ${fasta} > ${fasta_name}
    fi

    if [ "${is_gff_compressed}" == "true" ]; then
        gzip -c -d ${gff} > ${gff_name}
    fi

    run_dbcan \\
        easy_substrate \\
        ${args} \\
        --threads ${task.cpus} \\
        --db_dir dbcan_db \\
        --output_dir results \\
        --input_raw_data ${fasta_name} \\
        --gff_type prodigal \\
        --input_gff ${gff_name} \\
        --mode ${mode}

    # Bulk rename of the results, dbcan doesn't have a prefix parameter
    find results -type f | while read -r file; do
        mv "\$file" "\$(dirname "\$file")/${prefix}_\$(basename "\$file" | tr [:upper:] [:lower:])"
    done

    gzip results/*.*

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dbcan: \$(run_dbcan version | sed "s/dbCAN version: //")
        dbcan_db: "${db_version}"
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir results
    touch results/${prefix}_overview.tsv
    touch results/${prefix}_dbcan_hmm_results.tsv
    touch results/${prefix}_dbCANsub_hmm_results.tsv
    touch results/${prefix}_diamond.out

    gzip results/*

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        dbcan: \$(run_dbcan version | sed "s/dbCAN version: //")
        dbcan_db: ${db_version}
    END_VERSIONS
    """
}
