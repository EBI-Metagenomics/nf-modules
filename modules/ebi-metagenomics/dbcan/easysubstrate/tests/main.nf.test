nextflow_process {

    name "Test Process DBCAN"
    script "../main.nf"
    process "DBCAN"

    // dbCAN will fail if the db is in a folder with a . at some point
    // this is due to some bad coding to get a file extension
    // https://github.com/linnabrown/run_dbcan/blob/master/dbcan/cli/syntenic_plot.py#L431
    // so when running this test locally - use NFT_WORKDIR and set the nf-test to something different than .nf-test

    tag "modules"
    tag "dbcan"
    tag "modules_ebimetagenomics"
    tag "dbcan/easysubstrate"

    test("dbCAN with the built in subsampled test ecoli faa and gff in protein mode") {

        config "./nextflow.config"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.faa", checkIfExists: true), // protein faa

                ]
                input[1] = [
                    [ id:'test' ], // meta map
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.gff", checkIfExists: true) // protein gff
                ]
                input[2] = [
                    file("${moduleDir}/tests/dbcan_test_db", type: 'dir', checkIfExists: true),
                    "5.1.1"
                ]
                input[3] = "protein"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                // The folder is empty now (?)
                { assert snapshot(
                    process.out.out,
                    process.out.versions.collect{ path(it).yaml },
                ).match() }
            )
        }
    }

    test("dbCAN with the built in subsampled test ecoli faa and gbk - stub") {

        options "-stub"

        config "./nextflow.config"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.faa", checkIfExists: true), // protein faa

                ]
                input[1] = [
                    [ id:'test' ], // meta map
                    file("${moduleDir}/tests/dbcan_test_db/ecoliK12MG1655_test.gff", checkIfExists: true) // protein gff
                ]
                input[2] = [
                    file("${moduleDir}/tests/dbcan_test_db", type: 'dir', checkIfExists: true),
                    "5.1.1_V1"
                ]
                input[3] = "protein"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    process.out.versions.collect{ path(it).yaml }
                ).match() }
            )
        }
    }
}
