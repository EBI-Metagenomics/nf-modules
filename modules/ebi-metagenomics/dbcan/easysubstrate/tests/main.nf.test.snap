{
    "dbCAN with the built in subsampled test ecoli faa and gff in protein mode": {
        "content": [
            null,
            [
                {
                    "DBCAN": {
                        "dbcan": "5.1.2",
                        "dbcan_db": "5.1.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T12:18:01.428266"
    },
    "dbCAN with the built in subsampled test ecoli faa and gbk - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_overview.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dbcan_hmm_results.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dbcansub_hmm_results.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_diamond.out.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    "versions.yml:md5,cbaee965c3f3816dc227d352f6095148"
                ],
                "cgc_gff": [
                    
                ],
                "cgc_standard_tsv": [
                    
                ],
                "dbhmm_output_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dbcan_hmm_results.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "dbsub_output_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dbcansub_hmm_results.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "diamond_output": [
                    [
                        {
                            "id": "test"
                        },
                        "test_diamond.out.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "overview_txt": [
                    [
                        {
                            "id": "test"
                        },
                        "test_overview.tsv.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "substrate_prediction_tsv": [
                    
                ],
                "synteny_pdfs": [
                    
                ],
                "uniinput_faa": [
                    
                ],
                "versions": [
                    "versions.yml:md5,cbaee965c3f3816dc227d352f6095148"
                ]
            },
            [
                {
                    "DBCAN": {
                        "dbcan": "5.1.2",
                        "dbcan_db": "5.1.1_V1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T14:50:28.800518"
    }
}